pkg/common/consumer/consumer_manager.go

```diff
@@ -246,64 +278,123 @@ func (m *kafkaConsumerGroupManagerImpl) consume(ctx context.Context, groupId str
+	}
+
+	// The managedGroup's start channel must be created (that is, the managedGroup must be marked
+	// as "stopped") before closing the internal ConsumerGroup, otherwise the consume function would
+	// return control to the factory.
+	managedGrp.createRestartChannel()
+	// Close the inner sarama ConsumerGroup, which will cause our consume() function to stop
+	// and wait for the managedGroup to start again.
+	err := managedGrp.group.Close()
+	if err != nil {
+		groupLogger.Error("Failed To Close Managed ConsumerGroup", zap.Error(err))
+		// Don't leave the start channel open if the group.Close() call failed; that would be misleading
+		managedGrp.closeRestartChannel()
+		return err
 	}
-	groupLogger.Info("ConsumerGroup Not Managed - Ignoring Stop Request")
-	return fmt.Errorf("stop requested for consumer group not in managed list: %s", groupId)
+
+	return m.processLock(lock, groupId, false)
 }
 
 // startConsumerGroups creates a new Consumer Group based on the groupId provided
-func (m *kafkaConsumerGroupManagerImpl) startConsumerGroup(groupId string) error {
+func (m *kafkaConsumerGroupManagerImpl) startConsumerGroup(lock *commands.CommandLock, groupId string) error {
+	if err := m.processLock(lock, groupId, true); err != nil {
+		return err
+	}
+
 	groupLogger := m.logger.With(zap.String("GroupId", groupId))
 	groupLogger.Info("Starting Managed ConsumerGroup")
-	m.groupLock.Lock() // Don't allow m.groups to be modified while processing a start request
-	defer m.groupLock.Unlock()
-	if _, ok := m.groups[groupId]; ok {
-		group, err := m.factory.createConsumerGroup(groupId)
-		if err != nil {
-			groupLogger.Error("Failed To Restart Managed ConsumerGroup", zap.Error(err))
-			return err
+	if !m.IsManaged(groupId) {
+		groupLogger.Info("ConsumerGroup Not Managed - Ignoring Start Request")
+		return fmt.Errorf("start requested for consumer group not in managed list: %s", groupId)
+	}
+
+	group, err := m.factory.createConsumerGroup(groupId)
+	if err != nil {
+		groupLogger.Error("Failed To Restart Managed ConsumerGroup", zap.Error(err))
+		return err
+	}
+
+	m.groupLock.Lock()
+	m.groups[groupId].group = group
+	m.groups[groupId].closeRestartChannel() // Closing this allows the waitForStart function to finish
+	m.groupLock.Unlock()
+
+	return m.processLock(lock, groupId, false)
+}
+
+// getGroup returns a group from the groups map using the groupLock mutex
+func (m *kafkaConsumerGroupManagerImpl) getGroup(groupId string) *managedGroup {
+	m.groupLock.RLock()
+	defer m.groupLock.RUnlock()
+	return m.groups[groupId]
+}
+
+// processLock handles setting and removing the managedGroup's lock status.
+// For the managedGroup with the given groupId, if that group exists and the provided CommandLock fields warrant it,
+// this function will:
+// - if "before" is true, set the lockedBy field of the group
+// - if "before" is false, remove the lockedBy field of the group
```

Comment:
not entirely true - if "before" is false AND lock indicates `unlockafter` right ?

Reply:
That's part of "and the CommandLock field warrants it" -- but I will clean up this comment; it's confusing